contracts/infrastructure/dapp/ParaswapFilter.sol

```diff
@@ -33,55 +35,208 @@ interface IParaswap {
+    function isValid(address _wallet, address /*_spender*/, address _to, bytes calldata _data) external view override returns (bool) {
         // disable ETH transfer
         if (_data.length < 4) {
             return false;
         }
-        return getMethod(_data) == MULTISWAP;
+        bytes4 methodId = getMethod(_data);
+        if(methodId == MULTISWAP) {
+            return isValidMultiSwap(_wallet, _data);
+        } 
+        if(methodId == SIMPLESWAP) {
+            return isValidSimpleSwap(_wallet, _to, _data);
+        }
+        if(methodId == SWAP_ON_UNI) {
+            return isValidUniSwap(_to);
+        }
+        if(methodId == SWAP_ON_UNI_FORK) {
+            return isValidUniForkSwap(_to, _data);
+        }
+        if(methodId == MEGASWAP) {
+            return isValidMegaSwap(_wallet, _data);
+        }
+        return false;
+    }
+
+    function isValidMultiSwap(address _wallet, bytes calldata _data) internal view returns (bool) {
+        (IParaswap.SellData memory sell) = abi.decode(_data[4:], (IParaswap.SellData));
+        return hasValidBeneficiary(_wallet, sell.beneficiary) &&
+            hasTradableToken(sell.path[sell.path.length - 1].to) && 
+            hasValidExchangeAdapters(sell.path);
+    }
+
+    function isValidSimpleSwap(address _wallet, address _augustus, bytes calldata _data) internal view returns (bool) {
+        (, address[] memory callees,, uint256[] memory startIndexes,, address beneficiary) 
+            = abi.decode(_data[4:], (uint256[5],address[],bytes,uint256[],uint256[],address));
+        return hasValidBeneficiary(_wallet, beneficiary) && hasAuthorisedCallees(_augustus, callees, startIndexes, _data);
+    }
+
+    function isValidUniSwap(address _augustus) internal view returns (bool) {
+        return uniswapProxy == IParaswap(_augustus).getUniswapProxy();
+    }
+
+    function isValidUniForkSwap(address _augustus, bytes calldata _data) internal view returns (bool) {
+        if(uniswapProxy != IParaswap(_augustus).getUniswapProxy()) {
+            return false;
+        }
+
+        (address factory, bytes32 initCode) = abi.decode(_data[4:], (address, bytes32));
+
+        return factory != address(0) && initCode != bytes32(0) && (
+            (factory == uniForkFactory1 && initCode == uniForkInitCode1) ||
+            (factory == uniForkFactory2 && initCode == uniForkInitCode2) ||
+            (factory == uniForkFactory3 && initCode == uniForkInitCode3)
+        );
+    }
+
+    function isValidMegaSwap(address _wallet, bytes calldata _data) internal view returns (bool) {
+        (IParaswap.MegaSwapSellData memory sell) = abi.decode(_data[4:], (IParaswap.MegaSwapSellData));
+        return hasValidBeneficiary(_wallet, sell.beneficiary) &&
+            hasTradableToken(sell.path[0].path[sell.path[0].path.length - 1].to) && 
+            hasValidMegaPath(sell.path);
+    }
+
+    function hasAuthorisedCallees(
```

Comment:
I'm concerned about the gas overhead of calling the DapRegistry for every callee. What's the benefit of this approach over having the callee information locally? In both cases adding a new callee will require a filter addition/update thus take 7 days.

Reply:
It avoids duplication of the registry logic and content. 

`simpleSwap` offers almost no extra security over directly doing the trade via a `multiCall` and can just be thought of as a light wrapper over `multiCall`. So I think `AugustusSwapper` should be treated like any other wallet with regard to the `DappRegistry`. 

Ultimately a `simpleSwap` will cost one extra call to the `DappRegistry` compared to an equivalent `multiCall`. We could at some point decide to have the client deconstruct the subcalls of a `simpleSwap` and make them become subcalls of a pure `multiCall` instead (bypassing AugustusSwapper) to save the extra `DappRegistry` call.

Reply:
That being said, we could add a method to the DappRegistry to check an array of (spend, to, data) tuples.